net/tstun/linkattrs_linux.go

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) 2021 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package tstun
+
+import (
+	"fmt"
+	"unsafe"
+
+	"golang.zx2c4.com/wireguard/tun"
+
+	"golang.org/x/sys/unix"
+)
+
+// ethtoolCmd is struct ethtool_cmd from linux/ethtool.h
+type ethtoolCmd struct {
+	cmd     uint32
+	_       [8]byte
+	speedLo uint16
+	_       [14]byte
+	speedHi uint16
+	_       [34]byte
+}
+
+// ifreq is struct ifreq from linux/if.h.
+// We use this instead of unix.Ifreq because the latter
```

Comment:
We can just modify x/sys/unix to add the bits you need for both this and probably the `ethtool_cmd` above. (cc @mdlayher @tklauser who are very active maintainers of x/sys/unix)

I'd rather we do that over maintaining this copy.

Reply:
Actually, there's a TODO from @mdlayher already:

```go
// TODO(mdlayher): export as IfreqData? For now we can provide helpers such as
// IoctlGetEthtoolDrvinfo which use these APIs under the hood.

// An ifreqData is an Ifreq which carries pointer data. To produce an ifreqData,
// use the Ifreq.withData method.
type ifreqData struct {
        name [IFNAMSIZ]byte
        // A type separate from ifreq is required in order to comply with the
        // unsafe.Pointer rules since the "pointer-ness" of data would not be
        // preserved if it were cast into the byte array of a raw ifreq.
        data unsafe.Pointer
        // Pad to the same size as ifreq.
        _ [len(ifreq{}.Ifru) - SizeofPtr]byte
}

// withData produces an ifreqData with the pointer p set for ioctls which require
// arbitrary pointer data.
func (ifr Ifreq) withData(p unsafe.Pointer) ifreqData {
        return ifreqData{
                name: ifr.raw.Ifrn,
                data: p,
        }
}
```

Reply:
In addition to the above, it may be doable to use the ethtool netlink interface instead for some of these interactions. Though that one is fairly new and frankly I am not sure what kernels you all support for Tailscale.

https://github.com/mdlayher/ethtool

PRs certainly welcome.

Reply:
Been there, gotta love that. +1 to a new ifreq ioctl wrapper like the one for EthtoolDrvinfo though assuming genetlink is not amenable due to kernel version support.

Reply:
This doesn't affect iOS so 1.5K is fine.

Reply:
Cool. As we know Linux won't break userspace, so if adding a new ioctl to x/sys feels less intrusive then feel free rather than pulling my package. I just figured I'd offer another option.